### PR TITLE
statistics: reduce allocation of types.Context

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/52342

Problem Summary:

We still have https://github.com/pingcap/tidb/issues/52429, but we can reduce the allocations first.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/bootstrap.go [hunk 1]
```diff
@@ -524,7 +523,10 @@ func (h *Handle) initStatsFMSketch(cache statstypes.StatsCache) error {
 	return nil
 }
 
-func (*Handle) initStatsBuckets4Chunk(cache statstypes.StatsCache, iter *chunk.Iterator4Chunk) {
+func (_ *Handle) initStatsBuckets4Chunk(sctx sessionctx.Context, cache statstypes.StatsCache, iter *chunk.Iterator4Chunk) {
```
Comment (hawkingrei) [inline pkg/statistics/handle/bootstrap.go:526]:
```suggestion
func (*Handle) initStatsBuckets4Chunk(sctx sessionctx.Context, cache statstypes.StatsCache, iter *chunk.Iterator4Chunk) {
```

File: pkg/statistics/handle/handle.go [hunk 2]
```diff
@@ -122,6 +122,10 @@ func NewHandle(
 	handle.StatsGC = storage.NewStatsGC(handle)
 	handle.StatsReadWriter = storage.NewStatsReadWriter(handle)
 
+	// Invalid date values may be inserted into table under some relaxed sql mode. Those values may exist in statistics.
+	// Hence, when reading statistics, we should skip invalid date check. See #39336.
+	sc := initStatsCtx.GetSessionVars().StmtCtx
+	sc.SetTypeFlags(sc.TypeFlags().WithIgnoreInvalidDateErr(true).WithIgnoreZeroInDate(true))
```
Comment (hawkingrei) [inline pkg/statistics/handle/handle.go:128]: This ```sc``` is not used in the ```handle```.

**Review by hawkingrei — APPROVED** [review]

File: pkg/domain/domain.go [hunk 3]
```diff
@@ -2264,6 +2264,10 @@ func (do *Domain) StartLoadStatsSubWorkers(ctxList []sessionctx.Context) {
 		// We need to assign high priority to it so that we can get the stats as quick as we can.
 		ctx.GetSessionVars().StmtCtx.Priority = mysql.HighPriority
 		do.wg.Add(1)
+		// Invalid date values may be inserted into table under some relaxed sql mode. Those values may exist in statistics.
+		// Hence, when reading statistics, we should skip invalid date check. See #39336.
+		sc := ctx.GetSessionVars().StmtCtx
+		sc.SetTypeFlags(sc.TypeFlags().WithIgnoreInvalidDateErr(true).WithIgnoreZeroInDate(true))
```
Comment (hi-rustin) [inline pkg/domain/domain.go:2270]: Could we use a function for it? Then we don't need to copy the comments and code again in other places.

File: pkg/session/session.go [hunk 4]
```diff
@@ -1838,6 +1838,7 @@ func (s *session) useCurrentSession(execOption sqlexec.ExecOption) (*session, fu
 	prevSQL := s.sessionVars.StmtCtx.OriginalSQL
 	prevStmtType := s.sessionVars.StmtCtx.StmtType
 	prevTables := s.sessionVars.StmtCtx.Tables
+	prevTypeFlags := s.sessionVars.StmtCtx.TypeFlags()
```
Comment (hi-rustin) [inline pkg/session/session.go:1841]: Why do we need this change? Could you explain a little bit?

File: pkg/statistics/histogram.go [hunk 5]
```diff
@@ -1126,51 +1124,6 @@ func GetIndexPrefixLens(data []byte, numCols int) (prefixLens []int, err error)
 	return prefixLens, nil
 }
 
-// ExtractTopN extracts topn from histogram.
```
Comment (hi-rustin) [inline pkg/statistics/histogram.go:1129]: Good catch!

**Review by hi-rustin — COMMENTED** [review]
Do you have any test results? Could you please put it on GitHub?

**Comment by winoros** [comment]
> Do you have any test results? Could you please put it on GitHub?

The test env is under rebuilding(

File: pkg/domain/domain.go (hunk 3 above)
Comment (winoros) [reply pkg/domain/domain.go:2270]: Do you have any good ideas?
I'm not sure how to make it better(

File: pkg/session/session.go (hunk 4 above)
Comment (winoros) [reply pkg/session/session.go:1841]: The `useCurrentSession` will return a func as the `clean` func after the session executes a SQL.
We need to make sure that the CLEAN operation will not clean the flag we set.

**Review by time-and-fate — APPROVED** [review]
Please note that not all places use flags with `FlagIgnoreInvalidDateErr` and `FlagIgnoreZeroInDateErr`.
Others LGTM.

**Pull request merged**